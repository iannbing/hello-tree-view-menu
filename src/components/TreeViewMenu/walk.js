import React from 'react';

import ListGroupItem from './ListGroupItem';

const walk = (data, props) =>
  Object.entries(data)
    .sort((a, b) => a[1].index - b[1].index)
    .reduce(
      (all, [nodeName, node]) => [
        ...all,
        ...(node.key ? generateBranch(node, nodeName, props) : [])
      ],
      []
    );

const generateBranch = (node, nodeName, props) => {
  const { parent = '', level = 0, openNodes, searchTerm, activeKey } = props;
  const { nodes, key } = node;
  const currentNode = [parent, nodeName].filter(x => x).join('/');
  const isActive = activeKey === key;
  const isOpen = !!nodes && (openNodes.includes(currentNode) || !!searchTerm);

  const currentItem = getComponent({
    isOpen,
    currentNode,
    isActive,
    ...props,
    ...node
  });
  const nextLevelItems = isOpen
    ? walk(nodes, { ...props, parent: currentNode, level: level + 1 })
    : [];

  return [currentItem, ...nextLevelItems].filter(x => x);
};

const getComponent = ({
  isActive,
  isOpen,
  nodes,
  key,
  level,
  currentNode,
  label,
  searchTerm,
  getOnClickItem,
  toggleIcon
}) => {
  const showCurrent =
    !searchTerm ||
    label.toLowerCase().includes(searchTerm.trim().toLowerCase());
  const onClickItem = getOnClickItem({ node: currentNode, label, key });

  //   console.log(isOpen);

  return (
    showCurrent && (
      <ListGroupItem
        hasSubItems={!!nodes}
        isOpen={isOpen}
        level={level}
        onClick={onClickItem}
        active={isActive}
        key={currentNode}
        toggleIcon={toggleIcon}
      >
        {label}
      </ListGroupItem>
    )
  );
};

export default walk;
