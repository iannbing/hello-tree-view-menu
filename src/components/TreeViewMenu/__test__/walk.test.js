import walk from '../walk';

const mockData = {
  atd: {
    label: 'ATS Guide',
    key: 'ats',
    index: 1, // ATS Guide should be after Release Notes
  },
  releasenotes: {
    label: 'Release Notes',
    key: 'releasenotes',
    index: 0, // Release Notes should be first
    nodes: {
      'desktop-modeler': {
        label: 'Desktop Modeler',
        key: 'releasenotes/desktop-modeler',
        index: 0,
        nodes: {
          7: {
            label: '7',
            key: 'releasenotes/desktop-modeler/7',
            index: 0,
            nodes: {
              '7.0': {
                label: '7.0',
                key: 'releasenotes/desktop-modeler/7.0',
                index: 0,
              },
            },
          },
        },
      },
    },
  },
};

describe('walk', () => {
  it('should transpose data to a desired shape', () => {
    const result = walk({ data: mockData, openNodes: [], searchTerm: '7' });

    const expected = [
      {
        nodePath: 'releasenotes/desktop-modeler/7',
        index: 0,
        isOpen: true,
        key: 'releasenotes/desktop-modeler/7',
        label: '7',
        level: 2,
        nodes: {
          '7.0': {
            index: 0,
            key: 'releasenotes/desktop-modeler/7.0',
            label: '7.0',
          },
        },
        openNodes: [],
        parent: 'releasenotes/desktop-modeler',
        searchTerm: '7',
      },
      {
        nodePath: 'releasenotes/desktop-modeler/7/7.0',
        index: 0,
        isOpen: false,
        key: 'releasenotes/desktop-modeler/7.0',
        label: '7.0',
        level: 3,
        openNodes: [],
        parent: 'releasenotes/desktop-modeler/7',
        searchTerm: '7',
      },
    ];

    expect(result).toEqual(expected);
  });
});
