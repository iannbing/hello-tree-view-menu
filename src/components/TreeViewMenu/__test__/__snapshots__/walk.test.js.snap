// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walk should only render the matching results that contains "7" 1`] = `
<div>
  <ListGroupItem
    active={false}
    hasSubItems={true}
    isOpen={true}
    key="releasenotes/desktop-modeler/7"
    level={2}
    onClick={Object {}}
  >
    7
  </ListGroupItem>
  <ListGroupItem
    active={false}
    hasSubItems={false}
    isOpen={false}
    key="releasenotes/desktop-modeler/7/7.0"
    level={3}
    onClick={Object {}}
  >
    7.0
  </ListGroupItem>
</div>
`;
